### PR TITLE
[pnnl/SHAD#22] Simple wrapper to allocate one instance per Locality.

### PR DESCRIPTION
**Pull request by mminutoli**
This pull request is a proposal to solve issue #22.  It will be used as a building block to implement reductions.

### CODE REVIEW

File: include/shad/data_structures/one_per_locality.h [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+template <typename T>
+class OnePerLocality : public AbstractDataStructure<OnePerLocality<T>> {
+ public:
+  using ObjectID = typename AbstractDataStructure<OnePerLocality<T>>::ObjectID;
+  using SharedPtr =
+      typename AbstractDataStructure<OnePerLocality<T>>::SharedPtr;
+
+  /// @brief Constructor.
+  template <typename... Args>
+  explicit OnePerLocality(ObjectID oid, Args... args)
```
Comment (VitoCastellana) [inline include/shad/data_structures/one_per_locality.h]: This should be protected. Also, the documentation for the create method should be added.

**Comment by mminutoli** [comment]
@VitoCastellana: I have updated the pull request to incorporate your comments.  I added a test storing and reading a different value on each locality.  It will help when we will go back to multi-locality systems.

File: include/shad/data_structures/one_per_locality.h [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  ///
+  /// Creates instances of a T object on each locality.
+  ///
+  /// @tparam Args The list of types needed to build an instance of type T
+  ///
+  /// @param args The parameter pack to be forwarded to T' constructor.
+  /// @return A shared pointer to the local OnePerlocality instance.
+#ifdef DOXYGEN_IS_RUNNING
+  template <typename... Args>
+  static ShadHashmapPtr Create(Args... args);
```
Comment (VitoCastellana) [inline include/shad/data_structures/one_per_locality.h]: :D Lovely cut & paste :-P

File: test/unit_tests/data_structures/one_per_locality_test.cc [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  shad::rt::executeOnAll(
+      [](const shad::OnePerLocality<int>::ObjectID& oid) {
+        auto localPtr = shad::OnePerLocality<int>::GetPtr(oid);
+        ASSERT_TRUE(localPtr != nullptr);
+        uint32_t expectedValue =
+            10 + static_cast<uint32_t>(shad::rt::thisLocality());
+        ASSERT_EQ(static_cast<int>(*localPtr), expectedValue);
+      },
+      anInt->GetGlobalID());
+}
```
Comment (VitoCastellana) [inline test/unit_tests/data_structures/one_per_locality_test.cc]: anInt is not Destroyed properly

Comment (mminutoli) [reply test/unit_tests/data_structures/one_per_locality_test.cc]: nice catch!

**Comment by mminutoli** [comment]
@VitoCastellana It should be good to be merged now.

**Pull request merged**